harness/determined/layers/_workload_sequencer.py

```diff
@@ -0,0 +1,507 @@
+        if before is None:
+            return True
+        smaller_is_better = self.env.experiment_config["searcher"]["smaller_is_better"]
+        return (now < before) if smaller_is_better else (now > before)
+
+    def validate(self, op: Optional[_searcher.SearcherOp]) -> WorkloadGenerator:
+        # Report a validation step is starting.
+        self.training.set_status("validating")
+
+        wkld = workload.Workload(
+            kind=workload.Workload.Kind.COMPUTE_VALIDATION_METRICS,
+            e_id=self._exp_id,
+            t_id=self._trial_id,
+            s_id=self.state.step_id,
+            num_batches=0,
+            total_batches_processed=self.state.total_batches,
+        )
+
+        response = yield from yield_and_await_response(wkld)
+
+        # Validation step is complete, process the result.
+
+        if isinstance(response, workload.InvalidHP):
+            self.training.report_early_exit(_training.EarlyExitReason.INVALID_HP)
+            raise ShouldExit()
+
+        metrics = response["metrics"]["validation_metrics"]
+
+        # Check that the validation metrics computed by the model code
+        # includes the metric used by the search method.
+        searcher_metric_name = self.env.experiment_config["searcher"]["metric"]
+        if searcher_metric_name not in metrics:
+            raise RuntimeError(
+                f"Search method is configured to use metric '{searcher_metric_name}' but model "
+                f"definition returned validation metrics {list(metrics.keys())}. The metric "
+                "used by the search method must be one of the validation "
+                "metrics returned by the model definition."
+            )
+
+        # Check that the searcher metric has a scalar value so that it can be compared for
+        # search purposes. Other metrics don't have to be scalars.
+        searcher_metric = metrics[searcher_metric_name]
+        if not tensorboard.metric_writers.util.is_numerical_scalar(searcher_metric):
+            raise RuntimeError(
+                f"Searcher validation metric '{searcher_metric_name}' returned "
+                f"a non-scalar value: {searcher_metric}"
+            )
+
+        if math.isnan(searcher_metric):
+            raise RuntimeError(
+                f"Searcher validation metric '{searcher_metric_name}' returned a NaN value"
+            )
+
+        # Report to the searcher API first, so we don't end up in a situation where we die between
+        # reporting to the metrics API and when we come back we refuse to repeat a validation, but
+        # we also don't have any validation metrics to report the the searcher API.
+        #
+        # A simpler solution here would be to execute in the following order (which would be
+        # suitable for most customers to implement on their own):
+        #   - validation
+        #   - report to metrics API
+        #   - report to searcher API
```

Comment:
Is that only possible because the searcher API returns the final operation multiple times?

This makes me further think that the searcher API should return None when there's no work to do.

Reply:
> This makes me further think that the searcher API should return None when there's no work to do.

Eh, I guess `completed=True` in the response is still sufficient.

But the AdvancedSearcher already guarantees that you don't ever get a searcher op that has already been completed.

Reply:
it really doesn't make much of a difference at all, but i usually prefer monotonic things, and, as is, the op for a trial is almost monotonic.. it never retracts information like how many units you needed (only retracts prev ops). this doesn't really come up but in more complex systems, retracting info vs monotonically changing the op makes it easier to reason about distributed consistency.